### DISCUSSION TITLE
Proper way to lazily initialize a module-level value from an environment variable

### DISCUSSION THREAD

**Discussion by mwchase** [koka-lang/koka#294] [Q&A]
My apologies if I mess up terminology or anything in here. There must be something I don't understand, but I don't know what it is.

I'm trying to write a small library that makes the values of particular environment variables easily accessible to the programs using it, and I'm running into trouble. Here is the code that I'm currently tinkering with to try to understand the issue:

```koka
import std/os/env

pub fun get-var( name : string ) : ndet string
  name.get-env.default("")

pub fun meta-variable( name : string ) : delayed<<ndet>, string>
  delay{ name.get-var }

pub val my-var : delayed<<ndet>, string> = "MY_VAR".meta-variable  // Annotation added to see if it would help; no obvious difference.

fun main() : <console,ndet> ()
  "MY_VAR".get-var.println  // Works
  // my-var.force.println  // Doesn't compile
```
If I swap around where the comments are in `main()`, compilation fails with
```
repro-env.kk(13,16): error: effects do not match
  context        :   my-var.force.println
  term           :                println
  inferred effect: ndet
  expected effect: <console,ndet|_e>
  because        : effect cannot be subsumed
```
If I compare the type information interactively, I get
```
> :t my-var
check  : interactive
delayed<ndet,string>

> :t my-var.force
check  : interactive
ndet string

> :t "MY_VAR".get-var
check  : interactive
ndet string
```
So, how come passing an `ndet string` to `println` only works sometimes? What is round-tripping it through `delay` and `force` doing?

**Comment by mwchase** [comment]
I just now came back to this, and it looks like the problem is with the `: delayed<<ndet>, string>` annotation. I don't know if that's a supertype of what it should be, or if something is just strange in the compiler, but at least I can make progress on this if I remember what I was doing.

**Comment by TimWhiting** [comment]
```
import std/os/env

pub fun get-var( name : string ) : ndet string
  name.get-env.default("")

pub fun meta-variable<e>( name : string ) : delayed<<ndet|e>, string>
  delay{ name.get-var }

pub val my-var : some<e> delayed<<ndet|e>, string> = "PATH".meta-variable  // Annotation added to see if it would help; no obvious difference.

fun main() : <console,ndet> ()
  // "MY_VAR".get-var.println  // Works
  my-var.force().println  // compiles!
```

Here is a working example. Tricky issue.
